querybook/server/datasources/query_execution.py

```diff
@@ -116,20 +123,65 @@ def get_query_execution(query_execution_id):
 
 @register("/query_execution/<int:query_execution_id>/", methods=["DELETE"])
 def cancel_query_execution(query_execution_id):
-    with DBSession() as session:
-        execution = logic.get_query_execution_by_id(query_execution_id, session=session)
-        verify_query_engine_permission(execution.engine_id, session=session)
-        execution_dict = execution.to_dict(True) if execution is not None else None
+    execution = logic.get_query_execution_by_id(query_execution_id)
+    api_assert(
+        execution is not None, f"Invalid query execution id {query_execution_id}"
+    )
 
-        requestor = current_user.id
-        api_assert(
-            requestor == execution_dict["uid"], "You can only cancel your own queries"
+    # Check if user has access to execution
+    verify_query_engine_permission(execution.engine_id)
+
+    # Check if the user is indeed the one who issued it
+    api_assert(current_user.id == execution.uid, "You can only cancel your own queries")
+
+    # Check if the execution is "RUNNING"
+    api_assert(
+        execution.status
+        in [
+            QueryExecutionStatus.INITIALIZED,
+            QueryExecutionStatus.RUNNING,
+            QueryExecutionStatus.DELIVERED,
+        ],
+        "Execution is already completed",
+    )
+
+    def cancel_query_and_notify():
+        statement_executions = execution.statement_executions
+        if len(statement_executions) > 0:
+            logic.update_statement_execution(
+                statement_executions[-1].id,
+                status=StatementExecutionStatus.CANCEL,
+                completed_at=datetime.utcnow(),
+            )
+
+        execution_dict = logic.update_query_execution(
+            query_execution_id,
+            status=QueryExecutionStatus.CANCEL,
+            completed_at=datetime.utcnow(),
+        ).to_dict()
+
+        socketio.emit(
+            "query_cancel",
+            execution_dict,
+            namespace=QUERY_EXECUTION_NAMESPACE,
+            room=query_execution_id,
         )
 
-        if execution_dict and "task_id" in execution_dict:
-            task = run_query_task.AsyncResult(execution_dict["task_id"])
-            if task is not None:
-                task.abort()
+    if not execution.task_id:
+        cancel_query_and_notify()
+        return
+
+    task = run_query_task.AsyncResult(execution.task_id)
+
+    if task.state == "PENDING":
```

Comment:
celery task has another `RECEIVED` state, which also can be revoked. would be nice to add this case as well.